edalize/libero.py

```diff
@@ -0,0 +1,136 @@
+                         'desc': 'Default HDL (e.g. "VERILOG"'}
+                    ]
+                    }
+
+    argtypes = ['vlogdefine', 'vlogparam']
+
+    tool_options_defaults = {
+        'speed': '-1',
+        'dievoltage': '1.0',
+        'range': 'IND',
+        'defiostd': 'LVCMOS 1.8V',
+        'hdl': 'VERILOG',
+    }
+
+    def _set_tool_options_defaults(self):
+        for key, default_value in self.tool_options_defaults.items():
+            if not key in self.tool_options:
+                logger.info("Set Libero tool option %s to default value %s"
+                            % (key, str(default_value)))
+                self.tool_options[key] = default_value
+
+    """ Initial setup of the class
+
+    This calls the parent constructor, but also identifies whether
+    the current system is using a Standard or Pro edition of Quartus.
+    """
+
+    def __init__(self, edam=None, work_root=None, eda_api=None):
+        if not edam:
+            edam = eda_api
+
+        super(Libero, self).__init__(edam, work_root)
+
+    """ Configuration is the first phase of the build
+
+    This writes the project TCL file. It first collects all
+    sources, IPs and constraints and then writes them to the TCL file along
+    with the build steps.
+    """
+
+    def configure_main(self):
+        self._set_tool_options_defaults()
+        (src_files, incdirs) = self._get_fileset_files(force_slash=True)
+        self.jinja_env.filters['src_file_filter'] = self.src_file_filter
+        self.jinja_env.filters['pdc_file_filter'] = self.pdc_file_filter
+
+        escaped_name = self.name.replace(".", "_")
+
+        template_vars = {
+            'name': escaped_name,
+            'src_files': src_files,
+            'incdirs': incdirs,
+            'tool_options': self.tool_options,
+            'toplevel': self.toplevel,
+            'generic': self.generic,
+            'prj_root': "../prj",
+            'op': "{",
+            'cl': "}"
+        }
+
+        # Render the TCL project file
+        self.render_template('libero-project.tcl.j2',
+                             escaped_name + '-project.tcl',
+                             template_vars)
+
+        # Render the TCL run file
+        self.render_template('libero-run.tcl.j2',
+                             escaped_name + '-run.tcl',
+                             template_vars)
+
+    def src_file_filter(self, f):
+        file_types = {
+            'PDC': '-io_pdc {',
+        }
+        _file_type = f.file_type.split('-')[0]
+        if _file_type in file_types:
+            return file_types[_file_type] + f.name
+        return ''
+
+    def pdc_file_filter(self, f):
+        file_types = {
+            'PDC': 'constraint/io/',
+        }
+        _file_type = f.file_type.split('-')[0]
+        if _file_type in file_types:
+            filename = f.name.split("/")[-1]
+            return file_types[_file_type] + filename
+        return ''
+
+    def build_main(self):
```

Comment:
Here you want to call libero twice, either directly or create a makefile in the configure step that does it.
First you call `libero SCRIPT:<name of project script>` and then `libero SCRIPT:<name of run script>` If you put these commands as the default target in a makefile, you can drop the build_main function, as the superclass will call make by default

Reply:
Shouldn't it be in `run_main` instead of `build_main`?
I ask because I usually build it all in one machine (a Mac for example) and then move files to another machine (Windows) to use on Libero.

Reply:
Also in my case, I'd use "fusesoc build" to only build the files but I get a deprecation message telling to use "run".
If I use "run", "libero" doesn't exists in this machine.

Reply:
The three stages are defined like this
1. configure: The EDAM file is converted to tool-specific project files
2. build: A binary is being built. For an FPGA flow this is the FPGA image. For a simulation flow this will be the compiled simulation model
3. run: Not all flows define the run phase. A simulator will run the simulation here. Some FPGA flows will program a board at this stage, but most just leave it empty. You can check out my [award-winning Edalize video](https://www.youtube.com/watch?v=HuRtkpZqB34) for more details :)
So in your case you would only do the configure step on the first computer. You can do that by running `fusesoc run --target=yourtarget --setup your_core`. Then you can copy the build directory to the computer where you have libero

Reply:
Great! Fixed on last push.